doc/ppx-for-plugin-authors.rst

```diff
@@ -359,3 +359,176 @@ whole ``metaquot`` extension point. E.g. you can write:
+.. _error_extensionf: https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib/Location/index.html#val-error_extensionf
+
+.. |pexp_extension| replace:: ``pexp_extension``
+.. _pexp_extension: https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib/Ast_builder/Default/index.html#val-pexp_extension
+
+.. |Ast_builder| replace:: ``Ast_builder``
+.. _Ast_builder: https://ocaml-ppx.github.io/ppxlib/ppxlib/Ppxlib/Ast_builder/index.html
+
+In order to give a nice user experience when reporting errors or failures in a ppx, it is necessary to include as much of the generated content as possible. Most IDE tools, such as Merlin, rely on the AST for their features, such as displaying type, jumping to definition or showing the list of errors.
+
+A common way to report an error is to throw an exception. However, this method interrupts the execution flow of the ppxlib driver without giving it an AST to produce, preventing Merlin to work.
+
+Instead, it is better to always return a valid AST, as complete as possible, but with "error extension nodes" at every place where valid code generation was impossible. Error extension nodes are special extension nodes ``[%ocaml.error error_message]``, which makes valid AST but are interpreted later as errors, for instance by the compiler or Merlin. As all extensions nodes, they can be put anywhere in the AST, from structure items to expressions or pattern.
+
+For instance, suppose a rewriter is supposed to define a new record type, but there is an error in the generation of the type of one field. In order to have the most complete AST as output, the rewriter can still define the type and all of its fields, putting an extension node in place of the type of the faulty field:
+
+..
+   - not include the type at all, outputting only:
+
+   .. code:: ocaml
+
+      [%ocaml.error "type long_record was not defined due to foo"]
+
+   - define the type, but not its implementation:
+
+   .. code:: ocaml 
+
+      type long_record = [%ocaml.error "type long_record was not implemented due to foo"]
+
+   - define and implement the type, but not its faulty field:
+
+.. code:: ocaml
+
+   type long_record = { field_1: int ; field_2: [%ocaml.error "field_2 could not be implemented due to foo"]}
+
+Ppxlib provides a function in its API to create error extension nodes: |error_extensionf|_. This function creates an extension node, which has then to be transformed in the right kind of node using functions such as for instance |pexp_extension|_ from the ``Default`` module of |Ast_builder|_.
+
+As the ppxlib driver get a valid AST, it can’t detect whether the ppx failed or not. This AST will be passed as input to the next rewriter, which means that further transformations might get AST containing error nodes. It is therefore important for whole file transormations authors to treat these with respect: for instance, to never dump some part of the AST without thinking twice about the risk of loosing errors, which could result in a very hard to debug state.
```

Comment:
I decided to remove this sentence, to be put in a section about whole file rewriters (yet to be written) which will emphasize that no nodes should dumped, as for instance it could contain error nodes (and then link to this part of the manual).